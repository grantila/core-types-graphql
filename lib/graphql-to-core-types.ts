import {
	NodeDocument,
	ArrayType,
	BooleanType,
	CoreTypesErrorMeta,
	IntegerType,
	NamedType,
	NodeType,
	NumberType,
	ObjectProperty,
	OrType,
	RefType,
	StringType,
	UnsupportedError,
	ConversionResult,
	throwRelatedError,
} from 'core-types'
import {
	DocumentNode,
	FieldDefinitionNode,
	ListTypeNode,
	NamedTypeNode,
	NameNode,
	UnionTypeDefinitionNode,
	parse,
	GraphQLError,
} from 'graphql'
import { parseDescription } from './annotation'
import { gqlStripRequired, isRequired } from './gql-utils'
import { GraphqlToCoreTypesOptions } from './types'
import { gqlLocationsToCoreTypesLocation } from './utils'


export function getGraphqlAst( source: string )
: DocumentNode
{
	return parse( source );
}

export function convertGraphqlToCoreTypes(
	source: string,
	options?: GraphqlToCoreTypesOptions,
)
: ConversionResult< NodeDocument >
{
	const {
		warn = ( message: string ) => console.warn( message ),
		unsupported = 'ignore',
	} = options ?? { };

	let document: DocumentNode;
	try
	{
		document = getGraphqlAst( source );
	}
	catch ( err )
	{
		if ( err instanceof GraphQLError )
			throwRelatedError(
				err,
				{
					source,
					loc: gqlLocationsToCoreTypesLocation(
						source,
						err.locations ?? [ ]
					)
				}
			);
		throw err;
	}

	if ( document.kind !== 'Document' )
		throw new UnsupportedError(
			`Invalid document type "${document.kind}"`,
			{
				blob: document,
				path: [ ],
				loc: document.loc,
			}
		);

	const notConvertedTypes: Array< string > = [ ];

	const types = document.definitions
	.map( ( definition ): NamedType< NodeType > | undefined =>
	{
		if ( definition.kind === 'UnionTypeDefinition' )
		{
			if ( !definition.types?.length )
			{
				notConvertedTypes.push( definition.name.value );
				return;
			}

			else if ( definition.types.length === 1 )
			{
				return {
					...parseCommonFields( definition ),
					...parseType( definition.types[ 0 ] ),
				};
			}

			else // Multiple types in union => or-type
				return {
					...parseCommonFields( definition ),
					type: 'or',
					or: definition.types.map( type => parseType( type ) ),
				};
		}
		else if ( definition.kind === 'ObjectTypeDefinition' )
		{
			const fields = definition.fields ?? [ ];
			return {
				...parseCommonFields( definition ),
				type: 'object',
				additionalProperties: false,
				properties: Object.fromEntries(
					fields.map( field => parseObjectField( field ) )
				)
			};
		}
		else
		{
			if ( unsupported === 'ignore' )
			{
				const name = ( definition as any )?.name?.value;
				if ( name )
					notConvertedTypes.push( name );
				return;
			}

			const message = `GraphQL kind ${definition.kind} not supported`;
			const meta: CoreTypesErrorMeta = {
				blob: definition,
				source,
				loc: definition.loc,
			};

			if ( unsupported === 'warn' )
				warn( message, meta );
			else
				throw new UnsupportedError( message, meta );
		}
	} )
	.filter( < T >( t: T ): t is NonNullable< T > => !!t );

	return {
		data: { version: 1, types },
		convertedTypes: types.map( ( { name } ) => name ),
		notConvertedTypes,
	};
}

function parseCommonFields(
	type: Pick< UnionTypeDefinitionNode, 'description' | 'loc' | 'name' >
)
: Omit< NamedType< NodeType >, 'type' >
{
	return {
		name: parseName( type.name ),
		...parseCommonFieldsWithoutName( type ),
	};
}

function parseCommonFieldsWithoutName(
	type: Pick< UnionTypeDefinitionNode, 'description' | 'loc' >
)
: Omit< NamedType< NodeType >, 'type' | 'name' >
{
	return {
		...parseDescription( type.description ),
		...( type.loc ? { loc: type.loc } : { } ),
	};
}

function parseName( name: NameNode ): string
{
	return name.value;
}

function parseFieldType( field: NamedTypeNode | ListTypeNode )
:
	| ArrayType
	| BooleanType
	| IntegerType
	| NumberType
	| StringType
	| RefType
	| OrType
{
	if ( field.kind === 'ListType' )
		return {
			type: 'array',
			elementType:
				isRequired( field.type )
				? parseFieldType( field.type.type )
				: {
					type: "or",
					or: [
						{ type: 'null' },
						parseFieldType( field.type ),
					]
				},
			...parseCommonFieldsWithoutName( field ),
		};
	else
		return parseType( field );
}

function parseObjectField( field: FieldDefinitionNode )
: [ string, ObjectProperty ]
{
	return [
		parseName( field.name ),
		{
			node: {
				...parseFieldType( gqlStripRequired( field.type ) ),
				...parseCommonFieldsWithoutName( field ),
			},
			required: isRequired( field.type ),
		}
	];
}

function parseType( type: NamedTypeNode )
: BooleanType | IntegerType | NumberType | StringType | RefType
{
	const name = parseName( type.name );
	if ( name === 'Boolean' )
		return { type: 'boolean' };
	else if ( name === 'Int' )
		return { type: 'integer' };
	else if ( name === 'Float' )
		return { type: 'number' };
	else if ( name === 'String' )
		return { type: 'string' };
	else
		return { type: 'ref', ref: name };
}
