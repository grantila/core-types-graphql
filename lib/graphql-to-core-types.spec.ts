import { convertGraphqlToCoreTypes } from './graphql-to-core-types'


describe( "graphql-to-core-types", ( ) =>
{
	it( "unions and objects", ( ) =>
	{
		const schema = `
		"Kind of a string"
		union Stringish = String

		"string or foo"
		union Bar = Stringish | Foo

		"""
		This is a foo

		fyi
		"""
		type Foo {
			"required prop"
			name: [Stringish!]!

			"""
			optional prop
			# example
				4
			# example
				{
					foo: 'bar'
				}
			# default
				42
			"""
			age: Int
			"The whiskey bar"
			bar: Bar!
		}`;
		const ct = convertGraphqlToCoreTypes( schema ).data.types;
		expect( ct ).toMatchSnapshot( );
	} );

	it( "enums", ( ) =>
	{
		const schema = `
		"Compression levels"
		enum CompressionLevel {
			None,
			Low,
			_2,
			_3,
			High,
		}`;
		const ct = convertGraphqlToCoreTypes( schema ).data.types;
		expect( ct ).toMatchSnapshot( );
	} );
} );
