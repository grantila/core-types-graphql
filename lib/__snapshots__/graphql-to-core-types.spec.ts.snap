// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-to-core-types enums 1`] = `
[
  {
    "description": "Compression levels",
    "enum": [
      "None",
      "Low",
      "_2",
      "_3",
      "High",
    ],
    "loc": {
      "end": 93,
      "start": 3,
    },
    "name": "CompressionLevel",
    "title": "CompressionLevel",
    "type": "string",
  },
]
`;

exports[`graphql-to-core-types unions and objects 1`] = `
[
  {
    "description": "Kind of a string",
    "loc": {
      "end": 48,
      "start": 3,
    },
    "name": "Stringish",
    "title": "Stringish",
    "type": "string",
  },
  {
    "description": "string or foo",
    "loc": {
      "end": 97,
      "start": 52,
    },
    "name": "Bar",
    "or": [
      {
        "ref": "Stringish",
        "type": "ref",
      },
      {
        "ref": "Foo",
        "type": "ref",
      },
    ],
    "title": "Bar",
    "type": "or",
  },
  {
    "additionalProperties": false,
    "description": "This is a foo

fyi",
    "loc": {
      "end": 350,
      "start": 101,
    },
    "name": "Foo",
    "properties": {
      "age": {
        "node": {
          "default": "42",
          "description": "optional prop",
          "examples": [
            "4",
            "{
	foo: 'bar'
}",
          ],
          "loc": {
            "end": 312,
            "start": 193,
          },
          "title": "Foo.age",
          "type": "integer",
        },
        "required": false,
      },
      "bar": {
        "node": {
          "description": "The whiskey bar",
          "loc": {
            "end": 346,
            "start": 316,
          },
          "ref": "Bar",
          "title": "Foo.bar",
          "type": "ref",
        },
        "required": true,
      },
      "name": {
        "node": {
          "description": "required prop",
          "elementType": {
            "ref": "Stringish",
            "type": "ref",
          },
          "loc": {
            "end": 188,
            "start": 150,
          },
          "title": "Foo.name",
          "type": "array",
        },
        "required": true,
      },
    },
    "title": "Foo",
    "type": "object",
  },
]
`;
